File: erlang/meetup/src/meetup.erl
-module(meetup).

-export([meetup/4]).

meetup(Year, Month, DayOfWeek, Week) ->
    FirstDay = week_first_day(Week, Year, Month),
    find_date(Year, Month, weekday(DayOfWeek), FirstDay).

week_first_day(first, _, _) -> 1;
week_first_day(second, _, _) -> 8;
week_first_day(third, _, _) -> 15;
week_first_day(fourth, _, _) -> 22;
week_first_day(teenth, _, _) -> 13;
week_first_day(last, Year, Month) ->
    calendar:last_day_of_the_month(Year, Month) - 6.

weekday(monday) -> 1;
weekday(tuesday) -> 2;
weekday(wednesday) -> 3;
weekday(thursday) -> 4;
weekday(friday) -> 5;
weekday(saturday) -> 6;
weekday(sunday) -> 7.

find_date(Year, Month, DayOfWeek, Day) ->
    case calendar:day_of_the_week(Year, Month, Day) of
        DayOfWeek -> {Year, Month, Day};
        _ -> find_date(Year, Month, DayOfWeek, Day + 1)
    end.

File: php/meetup/Meetup.php
<?php

declare(strict_types=1);

function meetup_day(int $year, int $month, string $which, string $weekday): DateTimeImmutable
{
    $day = starting_day($year, $month, $which);
    $date = new DateTime("$year-$month-$day");

    while (ucfirst($date->format('l')) !== $weekday) {
        $date->modify('+1 day');
    }

    return DateTimeImmutable::createFromMutable($date);
}

function starting_day(int $year, int $month, string $week): int
{
    $last = new DateTime("$year-$month-01");
    $last->modify('last day of this month');

    switch ($week) {
        case "first": return 1;
        case "second": return 8;
        case "third": return 15;
        case "fourth": return 22;
        case "last": return (int) $last->format('d') - 6;
        case "teenth": return 13;
    }
}
?>
